test/ruby/test_exception.rb

```diff
@@ -149,6 +149,41 @@ def self.meow
     end
   end
 
+  def test_exception_with_defined_in_short_circuit_if_condition
+    bug = '[ruby-core:20501]'
+    conds = [
+      "false && defined?(Some::CONSTANT)",
+      "true || defined?(Some::CONSTANT)",
+      "(false && defined?(Some::CONSTANT))",
+      "(true || defined?(Some::CONSTANT))",
+      "@val && false && defined?(Some::CONSTANT)",
+      "@val || true || defined?(Some::CONSTANT)"
+    ]
+
+    conds.each do |cond|
+      code = %Q{
+        def my_method
+          var = nil
+          if #{cond}
+            "here"
+          end
+          raise
+        end
+        begin
+          my_method
+        rescue
+          print 'ok'
+        else
+          print 'ng'
+        end
+      }
+      # invoke in subprocess because the bug caused a segfault
+      out, _err, status = EnvUtil.invoke_ruby('--disable-gems', code, true, false)
+      assert_predicate(status, :success?, bug)
+      assert_equal 'ok', out
```

Comment:
We have the dedicated method to run assertions in a separated process.

```suggestion
      assert_separately([], "#{<<~"begin;"}\n#{<<~'end;'}")
      begin;
        def my_method
          var = nil
          if #{cond}
            var = var = "here"
          end
          raise
        end
        assert_raise(RuntimeError) do
          my_method
        end
      end;
```

And this doesn't seem related to behaviors of exception, maybe test_syntax.rb?

Reply:
Ok I moved to test_syntax, but I just tried using assert_separately and couldn't reproduce the issue. I could reproduce it only in a clean environment with no test environment required, so that's why I didn't use it. I've added a comment about that in my most recent commit.